### PR TITLE
Put source in document

### PR DESCRIPTION
**Pull request by panglesd**
- Turning source into a document AST is done in document and not in html
- Compatshape also returns relevant info from cmt

### CODE REVIEW

File: doc/driver.md [hunk 1]
```diff
@@ -325,18 +341,43 @@ let compile_deps f =
...
+  in
+  match source_dir_of_odoc_lib lib with
+  | None -> None, None
+  | Some path ->
+      let path = Fpath.( / ) path filename in
+      let find_by_extension path exts =
+        List.map (fun ext -> Fpath.add_ext ext path) exts
+        |> List.find_opt (fun f -> Bos.OS.File.exists f |> get_ok)
+      in
+      let impl = find_by_extension path ["pp.ml" ;"ml-gen"; "ml"] in
```
Comment (Julow) [inline doc/driver.md:366]: `.pp.ml` is for the output of a ppx, right ? If yes, we definitely want to show the original source.

If you want to handle cppo, I think the preprocessed source is placed into just `.ml` (`.cppo.ml` seems to be the original source).

File: src/source_info/lines.ml [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  let is_line_directive line =
+    if String.length line > 0 then String.get line 0 = '#' else false
+  in
+  let _, poses, _ =
+    List.fold_left
+      (fun (i, poses, count) line ->
+        let l = String.length line in
+        let new_i, new_pos =
+          (* Ignore line directives *)
+          if is_line_directive line then (i, poses)
```
Comment (Julow) [inline src/source_info/lines.ml:12]: What is this doing ?
I think line directives are already counted in the locations. There's a directive for defining the line number (starting on the next line), perhaps that's what you meant to handle ?

```
# 31 "src/loader/cmi.ml"
```

File: doc/odoc.mld [hunk 3]
```diff
@@ -59,3 +59,7 @@ The main other pages of this site:
 - {!page-parent_child_spec} Parent/Child specification
 - {!page-interface} Interface guarantees
 
+{1 Internal API}
+
+{!modules: Odoc_html Odoc_document Odoc_latex Odoc_loader Odoc_manpage Odoc_model Odoc_model_desc Odoc}
```
Comment (Julow) [inline doc/odoc.mld:64]: There are similar links into contributing.mld already. This change is out of scope.

File: src/source_info/lines.ml (hunk 2 above)
Comment (panglesd) [reply src/source_info/lines.ml:12]: Yes, this code was ignoring line directives only to have a working demo. I included the code in this branch by mistake, sorry about that. Line directives need to be handled correctly (this is #15). @EmileTrotignon and I are working on this...

File: doc/driver.md (hunk 1 above)
Comment (panglesd) [reply doc/driver.md:366]: It is the output of a "string preprocessor".
It is the file that the ocaml compiler will consume and all locations will be relative to this file... (except if there are lines directives to the original file, in which case locations can maybe workish)
Here are what I see as advantages/disadvantages of using the `.pp.ml` versus using the `.ml`:
- Using the `.pp.ml`: will necessarily be of valid OCaml file, so syntax highlighting will work. Will work with Cppo and includes: we see the reconstructed file.
- Using the `.ml`: syntax highlighting will certainly fail. Will only workish if line directives are applied well. But, the user sees the same file as in the editor, which is good?

Comment (Julow) [reply doc/driver.md:366]: I see. You are right, we want to take the `.pp.ml`, which is what the locations will be based on. Syntax highlighting is less of a problem compared to locations, which must be perfectly in sync.

File: doc/odoc.mld (hunk 3 above)
Comment (panglesd) [reply doc/odoc.mld:64]: You are right, I added them mostly to be able to go more easily to the documentation for testing purpose. But it's out of scope, I'll remove this.

**Comment by Julow** [comment]
Still problems in the last commit.

**Comment by Julow** [comment]
Merged into render-source-files.
